File: 0200-number-of-islands/0200-number-of-islands.go
func numIslands(grid [][]byte) int {
    islands := 0
    for i := 0; i < len(grid); i++ {
        for j := 0; j < len(grid[i]); j++ {
            if string(grid[i][j]) == "1" {
                traverseIsland(i, j, grid)
                islands++
            }
        }
    }
    return islands
}

func traverseIsland(i, j int, grid [][]byte) {
    queue := [][]int{
        {i, j},
    }
    for len(queue) != 0 {
        cur := queue[len(queue)-1]
        queue = queue[:len(queue)-1]
        
        grid[cur[0]][cur[1]] = 0
        neighbors := getNeighbors(cur[0], cur[1], grid)
        for _, n := range neighbors {
            if string(grid[n[0]][n[1]]) == "1" {
                queue = append(queue, n)
            }
        }
    }
    return
}

func getNeighbors(i, j int, grid[][]byte) [][]int {
    neighbors := make([][]int, 0, 4)
    
    if 0 < i {
        neighbors = append(neighbors, []int{i-1, j})
    }
    if i < len(grid)-1 {
        neighbors = append(neighbors, []int{i+1, j})
    }
    if 0 < j {
        neighbors = append(neighbors, []int{i, j-1})
    }
    if j < len(grid[i])-1 {
        neighbors = append(neighbors, []int{i, j+1})
    }
    
    return neighbors
}

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        islands = 0
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == "1":
                    islands += 1
                    self.traverse_island(i, j, grid)
        return islands            
        
    def traverse_island(self, i: int, j: int, grid: List[List[str]]) -> None:
        queue = [(i, j)]
        
        while queue:
            pos_i, pos_j = queue.pop()
            grid[pos_i][pos_j] = "2"
            neighbors = [
                (pos_i+1, pos_j),
                (pos_i-1, pos_j),
                (pos_i, pos_j+1),
                (pos_i, pos_j-1),
            ]
            
            for n in neighbors:
                if 0 <= n[0] < len(grid) and 0 <= n[1] < len(grid[0]):
                    if grid[n[0]][n[1]] == "1":
                        queue.append(n)
        return                